core/tx_pool.go

```diff
@@ -675,14 +718,282 @@ func (pool *TxPool) validateTx(tx *types.Transaction, local bool) error {
+	}
+	return nil
+}
+
+// validateStakingTx checks the staking message based on the staking directive
+func (pool *TxPool) validateStakingTx(tx *staking.StakingTransaction) error {
+	// from address already validated
+	from, _ := types.PoolTransactionSender(pool.signer, tx)
+	b32, _ := hmyCommon.AddressToBech32(from)
+
+	switch tx.StakingType() {
+	case staking.DirectiveCreateValidator:
+		msg, err := staking.RLPDecodeStakeMsg(tx.Data(), staking.DirectiveCreateValidator)
+		if err != nil {
+			return err
+		}
+		stkMsg, ok := msg.(*staking.CreateValidator)
+		if !ok {
+			return ErrInvalidMsgForStakingDirective
+		}
+		if from != stkMsg.ValidatorAddress {
+			return errors.WithMessagef(ErrInvalidSender, "staking transaction sender is %s", b32)
+		}
+		return pool.validateCreateValidatorMsg(stkMsg)
+	case staking.DirectiveEditValidator:
+		msg, err := staking.RLPDecodeStakeMsg(tx.Data(), staking.DirectiveEditValidator)
+		if err != nil {
+			return err
+		}
+		stkMsg, ok := msg.(*staking.EditValidator)
+		if !ok {
+			return ErrInvalidMsgForStakingDirective
+		}
+		if from != stkMsg.ValidatorAddress {
+			return errors.WithMessagef(ErrInvalidSender, "staking transaction sender is %s", b32)
+		}
+		return pool.validateEditValidatorMsg(stkMsg)
+	case staking.DirectiveDelegate:
+		msg, err := staking.RLPDecodeStakeMsg(tx.Data(), staking.DirectiveDelegate)
+		if err != nil {
+			return err
+		}
+		stkMsg, ok := msg.(*staking.Delegate)
+		if !ok {
+			return ErrInvalidMsgForStakingDirective
+		}
+		if from != stkMsg.DelegatorAddress {
+			return errors.WithMessagef(ErrInvalidSender, "staking transaction sender is %s", b32)
+		}
+		return pool.validateDelegateMsg(stkMsg)
+	case staking.DirectiveUndelegate:
+		msg, err := staking.RLPDecodeStakeMsg(tx.Data(), staking.DirectiveUndelegate)
+		if err != nil {
+			return err
+		}
+		stkMsg, ok := msg.(*staking.Undelegate)
+		if !ok {
+			return ErrInvalidMsgForStakingDirective
+		}
+		if from != stkMsg.DelegatorAddress {
+			return errors.WithMessagef(ErrInvalidSender, "staking transaction sender is %s", b32)
+		}
+		return pool.validateUndelegateMsg(stkMsg)
+	case staking.DirectiveCollectRewards:
+		msg, err := staking.RLPDecodeStakeMsg(tx.Data(), staking.DirectiveUndelegate)
+		if err != nil {
+			return err
+		}
+		stkMsg, ok := msg.(*staking.CollectRewards)
+		if !ok {
+			return ErrInvalidMsgForStakingDirective
+		}
+		if from != stkMsg.DelegatorAddress {
+			return errors.WithMessagef(ErrInvalidSender, "staking transaction sender is %s", b32)
+		}
+		return pool.validateCollectRewardsMsg(stkMsg)
+	default:
+		return ErrUnknownStakingDirective
+	}
+}
+
+func (pool *TxPool) validateCreateValidatorMsg(stkMsg *staking.CreateValidator) error {
+	if pool.currentState.IsValidator(stkMsg.ValidatorAddress) {
+		return errValidatorExist
+	}
+
+	// create temp validator to do sanity check
+	validatorWrapper, err := staking.CreateValidatorFromNewMsg(stkMsg, pool.chain.CurrentBlock().Number())
```

Comment:
This few steps is a copy of the actual applyStakingTransaction, better to refactor them into a single function. Otherwise, there can be consistency issue in the future.

Reply:
refactored the staking tx check and tx pool now calls it